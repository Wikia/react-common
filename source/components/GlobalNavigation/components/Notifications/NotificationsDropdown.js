import React, { useCallback, useContext } from 'react';
import PropTypes from 'prop-types';

import Dropdown from '../../../Dropdown';
import Icon from '../../../Icon';

import NotificationsContext from '../../context/NotificationContext';

import ListContainer from './ListContainer';

const NotificationToggle = ({ chevron, unreadCount }) => (
    <React.Fragment>
        {
            Boolean(unreadCount) && (
                <span className="wds-global-navigation__notifications-counter">
                    {unreadCount > 99 ? '99+' : unreadCount}
                </span>
            )
        }
        <Icon name="bell" className="icon-bell" />
        {chevron}
    </React.Fragment>
);

NotificationToggle.propTypes = {
    chevron: PropTypes.node.isRequired,
    unreadCount: PropTypes.number.isRequired,
};

const NotificationsDropdown = ({ track }) => {
    const { unreadCount, loadFirstPage } = useContext(NotificationsContext);
    const onMouseEnter = useCallback(() => {
        track({
            label: 'open-menu',
            category: 'on-site-notifications',
            action: 'click',
            value: unreadCount,
        });

        loadFirstPage();
    }, [unreadCount, track, loadFirstPage]);

    return (
        <Dropdown
            className="wds-global-navigation__notifications-dropdown"
            toggleClassName="wds-global-navigation__dropdown-toggle"
            contentClassName="wds-notifications__dropdown-content"
            onMouseEnter={onMouseEnter}
            toggle={chevron => <NotificationToggle chevron={chevron} unreadCount={unreadCount} />}
            iconName="dropdown-tiny"
            hasShadow
            dropdownRightAligned
        >
            <ListContainer track={track} />
        </Dropdown>
    );
};

NotificationsDropdown.propTypes = {
    track: PropTypes.func.isRequired,
};

export default NotificationsDropdown;
