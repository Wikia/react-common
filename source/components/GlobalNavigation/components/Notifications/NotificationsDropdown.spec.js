import React, { useContext, useCallback } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme/build';
import merge from 'lodash/merge';

import NotificationsDropdown from './NotificationsDropdown';

const defaultProps = {
    track: () => null,
};

const mockDefaultContext = { loadFirstPage: () => null, unreadCount: 0 };

function renderComponent(props) {
    const renderer = new ShallowRenderer();
    const computedProps = merge({}, defaultProps, props);

    renderer.render(<NotificationsDropdown {...computedProps} />);

    return renderer.getRenderOutput();
}

// TODO: remove mocks when enzyme/react-test-renderer will fully support hooks
jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: jest.fn(() => mockDefaultContext),
    useCallback: jest.fn(),
}));

/* eslint-disable-next-line jest/no-hooks */
beforeEach(() => {
    jest.resetModules();
});

test('NotificationsDropdown renders correctly with default props', () => {
    expect(renderComponent()).toMatchSnapshot();
});

test('NotificationsDropdown calls track and loadFirstPage onMouseEnter', () => {
    const loadFirstPageMock = jest.fn();
    const trackMock = jest.fn();
    useContext.mockImplementation(() => ({ ...mockDefaultContext, loadFirstPage: loadFirstPageMock }));
    useCallback.mockImplementation(callback => event => callback(event));
    const output = renderComponent({ track: trackMock });
    const wrapper = shallow(<div>{output}</div>);

    wrapper.find('Dropdown').props().onMouseEnter();

    expect(loadFirstPageMock).toBeCalledWith();
    expect(trackMock).toBeCalledWith(expect.any(Object));
});

test('NotificationsDropdown renders toggle correctly for unread counter greater than 0', () => {
    useContext.mockImplementation(() => ({ ...mockDefaultContext, unreadCount: 10 }));

    const output = renderComponent();
    const wrapper = shallow(<div>{output}</div>);
    const Toggle = wrapper.find('Dropdown').props().toggle(<div />);

    expect(shallow(Toggle)).toMatchSnapshot();
});

test('NotificationsDropdown renders toggle correctly for unread counter greater than 99', () => {
    useContext.mockImplementation(() => ({ ...mockDefaultContext, unreadCount: 120 }));

    const output = renderComponent();
    const wrapper = shallow(<div>{output}</div>);
    const Toggle = wrapper.find('Dropdown').props().toggle(<div />);

    expect(shallow(Toggle)).toMatchSnapshot();
});
