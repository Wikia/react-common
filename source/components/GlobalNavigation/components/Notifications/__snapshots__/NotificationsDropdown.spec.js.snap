// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsDropdown renders correctly with default props 1`] = `
<Dropdown
  canFlip={false}
  className="wds-global-navigation__notifications-dropdown"
  contentClassName="wds-notifications__dropdown-content"
  contentScrollable={false}
  dropdownLeftAligned={false}
  dropdownRightAligned={true}
  hasDarkShadow={false}
  hasShadow={true}
  iconName="dropdown-tiny"
  isActive={false}
  isLevel2={false}
  isNotHoverable={false}
  isStickedToParent={false}
  noChevron={false}
  onClose={null}
  onMouseEnter={null}
  toggle={[Function]}
  toggleAttrs={Object {}}
  toggleClassName="wds-global-navigation__dropdown-toggle"
>
  <ListContainer
    header=""
    track={[Function]}
  />
</Dropdown>
`;

exports[`NotificationsDropdown renders toggle correctly for unread counter greater than 0 1`] = `
<Fragment>
  <span
    className="wds-global-navigation__notifications-counter"
  >
    10
  </span>
  <Icon
    className="icon-bell"
    name="bell"
    small={false}
    tiny={false}
  />
  <div />
</Fragment>
`;

exports[`NotificationsDropdown renders toggle correctly for unread counter greater than 99 1`] = `
<Fragment>
  <span
    className="wds-global-navigation__notifications-counter"
  >
    99+
  </span>
  <Icon
    className="icon-bell"
    name="bell"
    small={false}
    tiny={false}
  />
  <div />
</Fragment>
`;
